From 8623cbb668d0ad4168ce43cfd35155f886347c81 Mon Sep 17 00:00:00 2001
From: HuaWen CHEN <quic_huawenc@quicinc.com>
Date: Mon, 23 Jan 2023 10:47:41 +0800
Subject: [PATCH] netifd: fix issues on ifindex

ifindex issue:
1. after wlan device down and up, the ifindex may change
   but netifd init the ifindex value only once.
   now, set ifindex again whenever get DEV_EVENT_ADD event
2. after netifd get RTM_NEWLINK,
   it query ifindex but got 0 due race-condition.
   now, renew ifindex when found the ifindex is 0

Change-Id: I97b724bb9078f3f8e045b2e3092c73c7bd5c5643
Signed-off-by:quic_huawenc@quicinc.com
---
 bridge.c       | 16 ++++++++++++++++
 device.c       | 38 ++++++++++++++++++++++++++++++++++++--
 device.h       |  3 ++-
 system-linux.c | 29 ++++++++++++++++++++++++-----
 4 files changed, 78 insertions(+), 8 deletions(-)

diff --git a/bridge.c b/bridge.c
index 7e61b9d..a058092 100644
--- a/bridge.c
+++ b/bridge.c
@@ -575,6 +575,22 @@ bridge_member_cb(struct device_user *dep, enum device_event ev)
 	case DEV_EVENT_AUTH_UP:
 		if (!bst->dev.active)
 			break;
+		//the device ifindex maybe change when wlan down and then up
+		//netifd now set ifindex only once, here to re-fresh the ifindex, otherwise bridge addif will failure
+		device_set_ifindex(dev, system_if_resolve(dev));
+
+		//netifd may get RTM_NETLINK before ifindex is generated,
+		//if the ifindex not ready, here we wait a few time to refresh the ifindex
+		if(dev->ifindex == 0) {
+			int try_number=0;
+			retry:
+				D(SYSTEM, "bridge_member_cb: sleep 50ms to wait %s.ifindex ready\n", bm->dev.dev->ifname);
+				usleep(50*100);
+				device_set_ifindex(dev, system_if_resolve(dev));
+				try_number++;
+				if(try_number <=2 && dev->ifindex == 0)
+					goto retry;
+		}
 
 		if (bridge_enable_member(bm))
 			break;
diff --git a/device.c b/device.c
index b3d0e85..c719f6f 100644
--- a/device.c
+++ b/device.c
@@ -141,10 +141,27 @@ void device_stp_init(void)
 
 static int set_device_state(struct device *dev, bool state)
 {
+	int ret = 0;
+	int retry = 0;
+	D(DEVICE, "set_device_state: [%s]%d\n", dev->ifname, state);
 	if (state) {
 		/* Get ifindex for all devices being enabled so a valid  */
 		/* ifindex is in place avoiding possible race conditions */
 		device_set_ifindex(dev, system_if_resolve(dev));
+
+		//netifd may get RTM_NETLINK before ifindex is generated,
+		//if the ifindex not ready, here we wait a few time to refresh the ifindex
+		if(dev->ifindex == 0) {
+			int try_number=0;
+			retry:
+				D(DEVICE, "set_device_state: sleep 50ms to wait %s.ifindex ready\n", dev->ifname);
+				usleep(50*100);
+				device_set_ifindex(dev, system_if_resolve(dev));
+				try_number++;
+				if(try_number <=4 && dev->ifindex == 0)
+					goto retry;
+		}
+
 		if (!dev->ifindex)
 			return -1;
 
@@ -154,7 +171,16 @@ static int set_device_state(struct device *dev, bool state)
 		dev->orig_settings.flags &= dev->settings.flags;
 		system_if_apply_settings(dev, &dev->settings, dev->settings.flags);
 
-		system_if_up(dev);
+		ret = system_if_up(dev);
+		while(ret < 0 && retry <=4 )
+		{
+			D(DEVICE, "set_device_state: system_if_up failed with ret=%d,errno:%d, sleep 50ms to retry\n", ret, errno);
+			retry++;
+			usleep(50*100);
+			ret = system_if_up(dev);
+		}
+		D(DEVICE, "set_device_state: system_if_up result: %d, errno:%d\n", ret, errno);
+
 	} else {
 		system_if_down(dev);
 		system_if_apply_settings(dev, &dev->orig_settings, dev->orig_settings.flags);
@@ -763,7 +789,7 @@ device_set_auth_status(struct device *dev, bool value)
 	device_broadcast_event(dev, DEV_EVENT_LINK_UP);
 }
 
-void device_set_present(struct device *dev, bool state)
+void device_set_present_private(struct device *dev, bool state)
 {
 	if (dev->sys_present == state)
 		return;
@@ -1235,5 +1261,13 @@ void device_hotplug_event(const char *name, bool add)
 	if (!dev || dev->type != &simple_device_type)
 		return;
 
+	if(dev->config_iface != NULL && add)  {
+	  if(dev->config_iface->state != IFS_UP && dev->config_iface->available == true)
+	  {
+            netifd_log_message(L_WARNING, "race condition: device %s was available but receive new hotplug on message\n", name);
+	    device_set_present(dev,false);
+	  }
+	}
+
 	device_set_present(dev, add);
 }
diff --git a/device.h b/device.h
index 37f8c37..7beca9d 100644
--- a/device.h
+++ b/device.h
@@ -333,7 +333,8 @@ void device_add_user(struct device_user *dep, struct device *dev);
 void device_remove_user(struct device_user *dep);
 void device_broadcast_event(struct device *dev, enum device_event ev);
 
-void device_set_present(struct device *dev, bool state);
+void device_set_present_private(struct device *dev, bool state);
+#define device_set_present(dev, state) { netifd_log_message(L_DEBUG, "call device_set_present from %s,line %d\n",  __func__, __LINE__); device_set_present_private(dev, state);}
 void device_set_link(struct device *dev, bool state);
 void device_set_ifindex(struct device *dev, int ifindex);
 int device_set_ifname(struct device *dev, const char *name);
diff --git a/system-linux.c b/system-linux.c
index 0f13a99..a700d31 100644
--- a/system-linux.c
+++ b/system-linux.c
@@ -643,8 +643,9 @@ static int cb_rtnl_event(struct nl_msg *msg, void *arg)
 	if (!system_get_dev_sysfs("carrier", dev->ifname, buf, sizeof(buf)))
 		link_state = strtoul(buf, NULL, 0);
 
-	if (dev->type == &simple_device_type)
-		device_set_present(dev, true);
+	netifd_log_message(L_DEBUG, "cb_rtnl_event: iface %s link_state=%d", dev->ifname, link_state);
+//	if (dev->type == &simple_device_type)
+//		device_set_present(dev, true);
 
 	device_set_link(dev, link_state ? true : false);
 
@@ -659,7 +660,12 @@ handle_hotplug_msg(char *data, int size)
 	char *cur, *end, *sep;
 	int skip;
 	bool add;
+	bool isRndis=false;
 
+	if(strstr(data, "rndis")) {
+		isRndis = true;
+		netifd_log_message(L_DEBUG, "receive hotplug msg: %s\n", data);
+	}
 	if (!strncmp(data, "add@", 4) || !strncmp(data, "move@", 5))
 		add = true;
 	else if (!strncmp(data, "remove@", 7))
@@ -672,6 +678,9 @@ handle_hotplug_msg(char *data, int size)
 
 	for (cur = data + skip; cur < end; cur += skip) {
 		skip = strlen(cur) + 1;
+		if(isRndis) {
+			netifd_log_message(L_DEBUG, " detail data:%s\n", cur);
+		}
 
 		sep = strchr(cur, '=');
 		if (!sep)
@@ -694,9 +703,12 @@ handle_hotplug_msg(char *data, int size)
 	if (!subsystem || !interface)
 		return;
 
-	if (interface_old)
+	if (interface_old) {
+		netifd_log_message(L_DEBUG, "interface_old:%s\n", interface_old);
 		device_hotplug_event(interface_old, false);
+	}
 
+	netifd_log_message(L_DEBUG, "interface %s action %d\n", interface, add);
 	device_hotplug_event(interface, add);
 }
 
@@ -871,10 +883,17 @@ retry:
 	ret = 0;
 	oldbr = system_get_bridge(dev->ifname, dev_buf, sizeof(dev_buf));
 	if (!oldbr || strcmp(oldbr, bridge->ifname) != 0) {
+
+		//retry set-ifindex to resolve racecondition that device not get ifindex
+		if(dev->ifindex == 0) {
+			device_set_ifindex(dev, system_if_resolve(dev));
+			D(SYSTEM, "system_bridge_addif: try get ifindex(new:%d)", dev->ifindex);
+		}
+
 		ret = system_bridge_if(bridge->ifname, dev, SIOCBRADDIF, NULL);
 		tries++;
-		D(SYSTEM, "Failed to add device '%s' to bridge '%s' (tries=%d): %s\n",
-		  dev->ifname, bridge->ifname, tries, strerror(errno));
+		D(SYSTEM, "Failed to add device '%s'/(ifindex=%d) to bridge '%s' (tries=%d): %s\n",
+		  dev->ifname, dev->ifindex, bridge->ifname, tries, strerror(errno));
 		if (tries <= 3)
 			goto retry;
 	}
-- 
2.17.1

